Content.Server/Salvage/SalvageSystem.cs

```diff
@@ -0,0 +1,344 @@
+            pulledTransform.WorldRotation = spAngle;
+
+            Report("salvage-system-announcement-arrived", ("timeLeft", HoldTime.TotalSeconds));
+            return true;
+        }
+        private void Report(string messageKey) =>
+            _chatManager.DispatchStationAnnouncement(Loc.GetString(messageKey), Loc.GetString("salvage-system-announcement-source"));
+        private void Report(string messageKey, params (string, object)[] args) =>
+            _chatManager.DispatchStationAnnouncement(Loc.GetString(messageKey, args), Loc.GetString("salvage-system-announcement-source"));
+
+        private void Transition(SalvageMagnetComponent magnet, TimeSpan currentTime)
+        {
+            switch (magnet.MagnetState.StateType)
+            {
+                case MagnetStateType.Attaching:
+                    if (SpawnSalvage(magnet))
+                    {
+                        magnet.MagnetState = new MagnetState(MagnetStateType.Holding, currentTime + HoldTime);
+                    }
+                    else
+                    {
+                        magnet.MagnetState = new MagnetState(MagnetStateType.CoolingDown, currentTime + CooldownTime);
+                    }
+                    break;
+                case MagnetStateType.Holding:
+                    Report("salvage-system-announcement-losing", ("timeLeft", DetachingTime.TotalSeconds));
+                    magnet.MagnetState = new MagnetState(MagnetStateType.Detaching, currentTime + DetachingTime);
+                    break;
+                case MagnetStateType.Detaching:
+                    if (magnet.AttachedEntity.HasValue)
+                    {
+                        SafeDeleteSalvage(magnet.AttachedEntity.Value);
+                    }
+                    else
+                    {
+                        Logger.ErrorS("salvage", "Salvage detaching was expecting attached entity but it was null");
+                    }
+                    Report("salvage-system-announcement-lost");
+                    magnet.MagnetState = new MagnetState(MagnetStateType.CoolingDown, currentTime + CooldownTime);
+                    break;
+                case MagnetStateType.CoolingDown:
+                    magnet.MagnetState = MagnetState.Inactive;
+                    break;
+            }
+        }
+
+        public override void Update(float frameTime)
+        {
+            foreach (var gridIdAndState in _salvageGridStates)
+            {
+                var state = gridIdAndState.Value;
+                if (state.ActiveMagnets.Count == 0) continue;
+                var gridId = gridIdAndState.Key;
+                // Not handling the case where the salvage we spawned got paused
+                // They both need to be paused, or it doesn't make sense
+                if (_pauseManager.IsGridPaused(gridId)) continue;
+                state.CurrentTime += TimeSpan.FromSeconds(frameTime);
```

Comment:
Seems wasteful? Why not just accumulate frametime as a float? At the very least it's constantly making the timespan when it could do it once outside of the loop.

Reply:
Yeah, I want to keep the timespan, because it makes the rest of the code easier to work with, and not need conversion, but moving this outside the loop is a good idea.